File: library/src/main/java/com/apppoweron/circularrevealbutton/ButtonColorProvider.java
package com.apppoweron.circularrevealbutton;

import android.animation.ValueAnimator;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.RippleDrawable;
import android.os.Build;
import android.support.annotation.RequiresApi;
import android.text.TextUtils;
import android.widget.Button;

import java.lang.reflect.Field;

/**
 * Collection of button regarded color methods
 */
class ButtonColorProvider {

    private Button button;

    private String originalTextColorHex;

    private Integer originalColorId;

    ButtonColorProvider(Button button) {
        this.button = button;
    }

    private String getButtonTextColorHex() {
        if (TextUtils.isEmpty(originalTextColorHex)) {
            originalTextColorHex = Integer.toHexString(button.getTextColors().getDefaultColor()).substring(2); //Cutting off the first two characters, because it is ARGB
        }
        return originalTextColorHex;
    }

    int translateAnimatedValueIntoFadingColor(ValueAnimator valueAnimator) {
        return Color.parseColor(getFirstAlphaTag((Integer) valueAnimator.getAnimatedValue()) + getButtonTextColorHex());
    }

    private String getFirstAlphaTag(int value) {
        String result = Integer.toHexString(value);
        if (result.length() < 2) {
            result = "0" + result;
        }
        return "#" + result;
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    int getButtonOriginalColorId() {
        if (originalColorId == null) {
            if (button.getBackground() instanceof ColorDrawable) {
                ColorDrawable buttonColor = (ColorDrawable) button.getBackground();
                originalColorId = buttonColor.getColor();
            } else if (button.getBackground() instanceof RippleDrawable) {
                RippleDrawable buttonColor = (RippleDrawable) button.getBackground();
                Drawable.ConstantState state = buttonColor.getConstantState();
                try {
                    Field colorField = state.getClass().getDeclaredField("mColor");
                    colorField.setAccessible(true);
                    ColorStateList colorStateList = (ColorStateList) colorField.get(state);
                    originalColorId = colorStateList.getDefaultColor();
                } catch (NoSuchFieldException e) {
                    e.printStackTrace();
                } catch (IllegalAccessException e) {
                    e.printStackTrace();
                }
            }
        }
        return originalColorId;
    }

}

File: library/src/main/kotlin/com/apppoweron/circularrevealbutton/ButtonColorProvider.kt
package com.apppoweron.circularrevealbutton

import android.animation.ValueAnimator
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.RippleDrawable
import android.os.Build
import android.support.annotation.RequiresApi
import android.text.TextUtils
import android.widget.Button

/**
 * Collection of button regarded color methods
 */
internal class ButtonColorProvider(private val button: Button) {

    private var originalTextColorHex: String? = null

    private var originalColorId: Int? = null

    private//Cutting off the first two characters, because it is ARGB
    val buttonTextColorHex: String
        get() {
            if (TextUtils.isEmpty(originalTextColorHex)) {
                originalTextColorHex = Integer.toHexString(buttonTextColor).substring(2)
            }
            return originalTextColorHex.toString()
        }

    val buttonTextColor: Int
        get() = button.textColors.defaultColor

    val buttonOriginalColorId: Int?
        @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
        get() {
            if (originalColorId == null) {
                if (button.background is ColorDrawable) {
                    val buttonColor = button.background as ColorDrawable
                    originalColorId = buttonColor.color
                } else if (button.background is RippleDrawable) {
                    val buttonColor = button.background as RippleDrawable
                    val state = buttonColor.constantState
                    try {
                        val colorField = state!!.javaClass.getDeclaredField("mColor")
                        colorField.isAccessible = true
                        val colorStateList = colorField.get(state) as ColorStateList
                        originalColorId = colorStateList.defaultColor
                    } catch (e: NoSuchFieldException) {
                        e.printStackTrace()
                    } catch (e: IllegalAccessException) {
                        e.printStackTrace()
                    }

                }
            }
            return originalColorId
        }

    fun translateAnimatedValueIntoFadingColor(valueAnimator: ValueAnimator): Int {
        return Color.parseColor(getFirstAlphaTag(valueAnimator.animatedValue as Int) + buttonTextColorHex)
    }

    private fun getFirstAlphaTag(value: Int): String {
        var result = Integer.toHexString(value)
        if (result.length < 2) {
            result = "0$result"
        }
        return "#$result"
    }

}
